### PR TITLE
Core subset time in gap fix

### PR DESCRIPTION
**Pull request by agstephens**
@huard and @Zeitsperre , I invite you to review this PR on clisops. There are two main components:

1. Changes in `clisops.core.subset#check_start_end_dates` decorator:
  - see: https://github.com/roocs/clisops/commit/96aceeccf2937b2261546503c571c89423e65c8e
  - there was an issue that picking times _between_ valid time values would cause an error, I have changed this so that the `start_date` and `end_date` will be _nudged_ to the adjacent valid value.

2. I have restructured our `clisops.ops.subset#subset` function so that it now correctly calls `clisops.core.subset` functions before working out how to chunk/split datasets:
 - see: https://github.com/roocs/clisops/commit/19ddf49179080935f065a63e78d717580de63154

### CODE REVIEW

**Comment by agstephens** [comment]
All, I'll fix the problems picked up by Travis.

**Comment by Zeitsperre** [comment]
Don't worry about the Windows build. My suggestion for the formatting fix:
```
conda install -c conda-forge pre_commit
pre-commit install
pre-commit run --all-files
```
You'll only need to run this once. After that, changes will be automatically applied when you commit changes.

File: clisops/core/subset.py [hunk 1]
```diff
@@ -58,8 +62,19 @@ def func_checker(*args, **kwargs):
                 stacklevel=2,
             )
             kwargs["start_date"] = da.time.min().dt.strftime("%Y").values
+        except ValueError:
+            warnings.warn(
+                '"start_date" has been nudged to nearest valid time step in xarray object.',
+                UserWarning,
+                stacklevel=2,
+            )
+            nudged = da.time.sel(time=slice(kwargs["start_date"], None)).values[0]
```
Comment (huard) [inline clisops/core/subset.py:71]: I suggest to use nearest neighbour instead: http://xarray.pydata.org/en/stable/indexing.html#nearest-neighbor-lookups

**Review by huard — APPROVED** [review]

**Review by Zeitsperre — APPROVED** [review]
This looks great. Nice catch!

File: clisops/core/subset.py (hunk 1 above)
Comment (agstephens) [reply clisops/core/subset.py:71]: @huard, given that we are asking for a range, the nearest neighbour lookup might include values outside of the requested range. The proposed solution will look _up_ from the lower bound and _down_ from the upper bound.

**Comment by agstephens** [comment]
> Don't worry about the Windows build. My suggestion for the formatting fix:
> 
> ```
> conda install -c conda-forge pre_commit
> pre-commit install
> pre-commit run --all-files
> ```
> 
> You'll only need to run this once. After that, changes will be automatically applied when you commit changes.

Thanks @Zeitsperre, I've been needing to learn how to do that for a _long_ time..Much appreciated.

**Pull request merged**